docs/enhancements/fep-000-api-device-fleet.md

```diff
@@ -0,0 +1,632 @@
+### Device API
+
+The device API is divided into multiple methods, some user facing, and other device/facing,
+this design facilitates cleaner RBAC control of the endpoint access. The diagram is only
+intended to illustrate the flow of information between the user/system and the device.
+
+```mermaid
+block-beta
+columns 1
+  user(["User"])
+
+  space
+
+
+  block
+    frontEndpoint["/devices/{name}"]
+  end
+
+  user --> frontEndpoint
+  frontEndpoint --> user
+
+  space
+  block
+    metadata
+    devSpec["spec"]
+    devStatus["status"]
+  end
+
+  space
+
+  frontEndpoint --> metadata
+  frontEndpoint --> devSpec
+  frontEndpoint --> devStatus
+
+  block
+     renderedEndpoint["/devices/{name}/rendered"]
+     statusEndpoint["/devices/{name}/status"]
+  end
+
+  space
+
+  statusEndpoint --> devStatus
+  statusEndpoint --> devStatus
+  devSpec --> renderedEndpoint
+  renderedEndpoint --> Device
+  Device --> statusEndpoint
+
+
+  Device(["Device"])
+
+  style frontEndpoint fill:#969,stroke:#333,stroke-width:4px
+  style renderedEndpoint fill:#969,stroke:#333,stroke-width:4px
+  style statusEndpoint fill:#969,stroke:#333,stroke-width:4px
+  ```
+
+#### `/devices/{name}` User facing method
+
+This endpoint can be accessed by the user or external systems to
+manage individual devices, also internally from the fleet specifications
+for devices.
+
+The proposed changes include:
+
+* Reincorporating device.spec as a visible part through this frontend API.
+
+* The `device.spec.os.image` contains a reference to the image to be used for the device.
+
+* The `device.spec.templateVersion` is removed. An annotation will be used to track the latest template applied to the
+  device. (See the fleet API section for more details)
+
+* The `device.spec.{containers, systemd}` fields are moved under `device.spec.monitoring.{containers, systemd}`
+
+* The `device.spec.monitoring.containers` contains a list of container patterns to be watched on the device (as in the fleet spec)
+
+* The `device.spec.monitoring.systemd` contains a list of systemd service patterns to be watched on the device (as in the fleet spec)
+
+* The `device.spec.config` contains a list of zero or more config items.
+
+* The config items could be inline definition of config files (`InlineConfigProviderSpec`)
+
+* Some config items reference external configurations (`KubernetesSecretProviderSpec`, `GitConfigProviderSpec`)
+
+* `GitConfigProviderSpec` config types add a `mountPath` field for consistency, in the same way k8s secrets add such field.
+
+* External references (image or config items) must support floating tags (`latest`, `main`, `v1.0`), although
+  **when managed from a fleet, the floating tags will be frozen as hashes** on the device spec by the
+  fleet controller.
+
+* Some external items are actually non-versioned like the K8s secrets, so are inherently floating.
+
+* The `device.status.conditions` field can be updated by the `device-controller` to provide status information
```

Comment:
I'm not clear what we mean with `device-controller` here, but would like to ensure we're on the same page that there is nothing else on the write path of agent --> API server --> DB when it comes to status updates.

Only the agent knows and can set the device `.status` fields indicating the update phase it is in (preparing to update, updating, rebooting, success or rollback&failure), reports this to the API server who adds the "last seen" info. At that point we have all info to display the "device status" information in the UI/API.

Reply:
It is what we discussed 2 days ago with avishay on a meeting, that sometimes the device controller (updating the rendered spec) needs to set conditions visible through the API. (i.e. a source is not found now... ).

We have in logical terms a `fleet-controller` that updates device specs based on fleet definitions,
And we have a `device-controller` that makes sure that the rendered spec is up to date with the sources provided in the device spec (and in the case of standalone devices , we use floating tags too).

Reply:
Let me try to phrase it more clearly.